### DISCUSSION TITLE
Return an object next to an array when reading contract view

### DISCUSSION THREAD

**Discussion by deanpress** [wevm/viem#1315] [Idea / Feature Request]
Currently, if you read a contract with multiple return values or a struct, you just get an array. It'd be great if, next to an array, you can also  get values by their attributes as defined in the abi.

For example:
```js
// Calling this using contract.read.getValues():
function getValues() public view returns (uint256[2] values, uint256 anotherValue)

// Currently returns
[ [ 123, 456 ], 789 ]

// Could return
[ [ 123, 456 ], 789 ] { values: [123, 456], anotherValue: 789 }
```

The latter is backwards compatible (same way as ethers returns contract return values) and allows you to query values by their key name, rather than having to refer to the abi every time to ensure you get the right key number.

Edit:

To keep it serializable in Javascript you could have it like this:
```js
await contract.read.getValues({return: "object"}): // "return" is "array" by default, keeping it compatible with the current logic

// returns { values: [123, 456], anotherValue: 789 }
```

**Comment by jxom** [comment]
Please see: https://viem.sh/docs/faq.html#why-is-a-contract-function-return-type-returning-an-array-instead-of-an-object

The second example you provided is not serializable in JavaScript.

**Reply by deanpress** [reply]
What if you add an option like `contract.read.myValues({ parse: "object" })` to return an object, and keep it as an array by default?

**Reply by jmrossy** [reply]
@jxom thanks for the link! Is there a recommended utility for converting to an object when devs explicitly want that? Something like `decodeFunctionResult` for for this use case?
